README.md
<img src="https://raw.githubusercontent.com/mcaceresb/mcaceresb.github.io/develop/assets/icons/gtools-icon/gtools-icon-text.png" alt="Gtools" width="500px"/>

[Overview](#faster-stata-for-big-data)
| [Installation](#installation)
| [Examples](#examples)
| [Remarks](#remarks)
| [FAQs <img src="https://upload.wikimedia.org/wikipedia/commons/6/64/Icon_External_Link.png" width="13px"/>](https://gtools.readthedocs.io/en/latest/faqs/index.html)
| [Benchmarks <img src="https://upload.wikimedia.org/wikipedia/commons/6/64/Icon_External_Link.png" width="13px"/>](https://gtools.readthedocs.io/en/latest/benchmarks/index.html)
| [Compiling <img src="https://upload.wikimedia.org/wikipedia/commons/6/64/Icon_External_Link.png" width="13px"/>](https://gtools.readthedocs.io/en/latest/compiling/index.html)

Faster Stata for big data. This packages uses C plugins and hashes to
provide a massive speed improvements to common Stata commands, including:
reshape, collapse, xtile, tabstat, isid, egen, pctile, winsor, contract,
levelsof, duplicates, unique/distinct, and more.

![Beta Version](https://img.shields.io/badge/beta-v1.11.7-blue.svg?longCache=true&style=flat-square)
![Supported Platforms](https://img.shields.io/badge/platforms-linux--64%20%7C%20osx--64%20%7C%20win--64-blue.svg?longCache=true&style=flat-square)
[![github linux status](https://github.com/mcaceresb/stata-gtools/actions/workflows/linux.yml/badge.svg?branch=develop)](https://github.com/mcaceresb/stata-gtools/actions/workflows/linux.yml)
[![github osx status](https://github.com/mcaceresb/stata-gtools/actions/workflows/osx.yml/badge.svg?branch=develop)](https://github.com/mcaceresb/stata-gtools/actions/workflows/osx.yml)
[![Appveyor Build status](https://img.shields.io/appveyor/ci/mcaceresb/stata-gtools/develop.svg?longCache=true&style=flat-square&label=windows-cygwin)](https://ci.appveyor.com/project/mcaceresb/stata-gtools)

Faster Stata for Big Data
-------------------------

This package provides a fast implementation of various Stata commands
using hashes and C plugins. The syntax and purpose is largely analogous
to their Stata counterparts; for example, you can replace `collapse`
with `gcollapse`, `reshape` with `greshape`, and so on. For a
comprehensive list of differences (including some extra features!)
see the [remarks](#remarks) below; for details and examples see [the
official project page](https://gtools.readthedocs.io).

__*Quickstart*__

```stata
ssc install gtools
gtools, upgrade
```

Some [quick benchmarks](https://raw.githubusercontent.com/mcaceresb/stata-gtools/develop/docs/benchmarks/quick.do):

_**NOTE:**_ Stata 17 introduced massive speed improvements to [sort and collapse](https://www.stata.com/new-in-stata/faster-stata-speed-improvements/).
In the MP version, in particular with many cores available, the native
`collapse`  can be up to twice as fast. (YMMV; overall native collapses 
could still be slower in some use cases.)  `gcollapse` remains faster
in SE and older Stata versions.

<img
    src="https://raw.githubusercontent.com/mcaceresb/stata-gtools/develop/docs/benchmarks/quick.png#gh-light-mode-only"
    alt="Gtools quick benchmark"
    style="display:block;margin-left:auto;margin-right:auto"
    width="80%"/>

<img
    src="https://raw.githubusercontent.com/mcaceresb/stata-gtools/develop/docs/benchmarks/quickdark.png#gh-dark-mode-only"
    alt="Gtools quick benchmark"
    style="display:block;margin-left:auto;margin-right:auto"
    width="80%"/>

__*Gtools commands with a Stata equivalent*__

| Function     | Replaces    | Speedup (IC / MP)              | Unsupported             | Extras                                  |
| ------------ | ----------- | ------------------------------ | ----------------------- | --------------------------------------- |
| gcollapse    | collapse    | -0.5 to 2 (Stata 17+); 4 to 100 (Stata 16 and earlier)  || Quantiles, merge, labels, nunique, etc. |
| greshape     | reshape     |  4 to 20  / 4 to 15            | "advanced syntax"       | `fast`, spread/gather (tidyr equiv)     |
| gegen        | egen        |  9 to 26  / 4 to 9 (+,.)       | labels                  | Weights, quantiles, nunique, etc.       |
| gcontract    | contract    |  5 to 7   / 2.5 to 4           |                         |                                         |
| gisid        | isid        |  8 to 30  / 4 to 14            | `using`, `sort`         | `if`, `in`                              |
| glevelsof    | levelsof    |  3 to 13  / 2 to 7             |                         | Multiple variables, arbitrary levels    |
| gduplicates  | duplicates  |  8 to 16 / 3 to 10             |                         |                                         |
| gquantiles   | xtile       |  10 to 30 / 13 to 25 (-)       |                         | `by()`, various (see [usage](https://gtools.readthedocs.io/en/latest/usage/gquantiles)) |
|              | pctile      |  13 to 38 / 3 to 5 (-)         |                         | Ibid.                                   |
|              | \_pctile    |  25 to 40 / 3 to 5             |                         | Ibid.                                   |
| gstats tab   | tabstat     |  10 to 50 / 5 to 30 (-)        | See [remarks](#remarks) | various (see [usage](https://gtools.readthedocs.io/en/latest/usage/gstats_summarize)) |
| gstats sum   | sum, detail |  10 to 20 / 5 to 10            | See [remarks](#remarks) | various (see [usage](https://gtools.readthedocs.io/en/latest/usage/gstats_summarize)) |

<small>(+) The upper end of the speed improvements are for quantiles
(e.g. median, iqr, p90) and few groups. Weights have not been
benchmarked.</small>

<small>(.) Only gegen group was benchmarked rigorously.</small>

<small>(-) Benchmarks computed 10 quantiles. When computing a large
number of quantiles (e.g. thousands) `pctile` and `xtile` are prohibitively
slow due to the way they are written; in that case gquantiles is hundreds
or thousands of times faster, but this is an edge case.</small>

__*Extra commands*__

| Function            | Similar (SSC/SJ)         | Speedup (IC / MP)       | Notes                         |
| ------------------- | ------------------------ | ----------------------- | ----------------------------- |
| fasterxtile         | fastxtile                | 20 to 30 / 2.5 to 3.5   | Allows `by()`                 |
|                     | egenmisc (SSC) (-)       | 8 to 25 / 2.5 to 6      |                               |
|                     | astile (SSC) (-)         | 8 to 12 / 3.5 to 6      |                               |
| gstats hdfe         |                          | (.)                     | Allows weights, `by()`        |
| gstats winsor       | winsor2                  | 10 to 40 / 10 to 20     | Allows weights                |
| gunique             | unique                   | 4 to 26 / 4 to 12       |                               |
| gdistinct           | distinct                 | 4 to 26 / 4 to 12       | Also saves results in matrix  |
| gtop (gtoplevelsof) | groups, select()         | (+)                     | See table notes (+)           |
| gstats range        | rangestat                | 10 to 20 / 10 to 20     | Allows weights; no flex stats |
| gstats transform    |                          |                         | Various statistical functions |

<small>(-) `fastxtile` from egenmisc and `astile` were benchmarked against
`gquantiles, xtile` (`fasterxtile`) using `by()`.</small>

<small>(+) While similar to the user command 'groups' with the 'select'
option, gtoplevelsof does not really have an equivalent. It is several
dozen times faster than 'groups, select', but that command was not written
with the goal of gleaning the most common levels of a varlist. Rather, it
has a plethora of features and that one is somewhat incidental. As such, the
benchmark is not equivalent and `gtoplevelsof` does not attempt to implement
the features of 'groups'</small>

<small>(.) Other than the dated 'hdfe' command, I do not know of a stata
command that residualizes variables from a set of fixed effects. The
'hdfe' command, as far as I can tell, morphed into the 'reghdfe'
package; the latter, however, is a fully-functioning regression command,
while 'gstats hdfe' only residualizes a set of variables.</small>

__*Regression models*__

_**WARNING:**_ Regression models are in beta and are only intended as utilities
to compute coefficients and standard errors. I do not recommend their use in
production; various post-estimation commands and statistics are _not_ availabe.
(See `gstats hdfe` for residualizing variables net of fixed effects.)

| Function            | Model   | Similar                        |
| ------------------- | ------- | -----------------------------  |
| gregress            | OLS     | `regress`, `reghdfe`           |
| givregress          | 2SLS    | `ivregress 2sls`, `ivreghdfe`  |
| gglm                | IRLS    | `logit`, `poisson`, `ppmlhdfe` |

All commands allow the user to optionally add:

- `absorb()` for high-dimensional fixed effects absorptions.
- `cluster()` for clustering (multiple covariates assume clusters are nested).
- `by()` for regressions by group.
- `weights` for weighted versions. Unlike other weights, `fweights` are assumed to refer to the _number_ of observations.

Linear regression is computed via OLS (or WLS), IV regression is
computed via two-stage least squares (2SLS), and GLM (poisson or logit)
regression is computed via iteratively reweighted least squares (IRLS). 
See the [TODO](#todo) section for planned features, or the
[Missing Features](https://gtools.readthedocs.io/en/latest/usage/gregress/index.html#missing-features)
section in the documentation for what is missing before the first
non-beta release.

__*Extra features*__

Several commands offer additional features on top of the massive
speedup. See the [remarks](#remarks) section below for an overview; for
details and examples, see each command's help page:

- [gcollapse](https://gtools.readthedocs.io/en/latest/usage/gcollapse/index.html#examples)
- [greshape](https://gtools.readthedocs.io/en/latest/usage/greshape/index.html#examples)
- [gquantiles](https://gtools.readthedocs.io/en/latest/usage/gquantiles/index.html#examples)
- [gstats sum/tab](https://gtools.readthedocs.io/en/latest/usage/gstats_summarize/index.html#examples)
- [gstats transform/range/moving](https://gtools.readthedocs.io/en/latest/usage/gstats_transform/index.html#examples)
- [glevelsof](https://gtools.readthedocs.io/en/latest/usage/glevelsof/index.html#examples)
- [gtoplevelsof](https://gtools.readthedocs.io/en/latest/usage/gtoplevelsof/index.html#examples)
- [gegen](https://gtools.readthedocs.io/en/latest/usage/gegen/index.html#examples)
- [gdistinct](https://gtools.readthedocs.io/en/latest/usage/gdistinct/index.html#examples)
- [gregress](https://gtools.readthedocs.io/en/latest/usage/gregress/index.html#examples)
- [givregress](https://gtools.readthedocs.io/en/latest/usage/givregress/index.html#examples)
- [gglm](https://gtools.readthedocs.io/en/latest/usage/gglm/index.html#examples) (poisson and logit)

In addition, several commands take gsort-style input, that is

```stata
[+|-]varname [[+|-]varname ...]
```

This does not affect the results in most cases, just the sort order.
Commands that take this type of input include:

- gcollapse
- gcontract
- gegen
- glevelsof
- gtop (gtoplevelsof)

__*Ftools*__

The commands here are also faster than the commands provided by
`ftools`; further, `gtools` commands take a mix of string and numeric
variables, which is a limitation of `ftools`. (Note I could not get
several parts of `ftools` working on the Linux server where I have
access to Stata/MP; hence the IC benchmarks.)

| Gtools    | Ftools        | Speedup (IC) |
| --------- | ------------- | ------------ |
| gcollapse | fcollapse     | 2-9          |
| gegen     | fegen         | 2.5-4 (+)    |
| gisid     | fisid         | 4-14         |
| glevelsof | flevelsof     | 1.5-13       |
| hashsort  | fsort         | 2.5-4        |

<small>(+) Only egen group was benchmarked rigorously.</small>

__*Limitations*__

- `strL` variables only partially supported on Stata 14 and above;
  `gcollapse`, `gcontract`, and `greshape` do not support `strL` variabes.

- Due to a Stata bug, gtools cannot support more
  than `2^31-1` (2.1 billion) observations. See [this
  issue](https://github.com/mcaceresb/stata-gtools/issues/43)

- Due to limitations in the Stata Plugin Interface, gtools

changelog.md
Change Log
==========

## gtools-1.11.7 (2023-11-08)

### Bug fixes

- `greg` with `alphas()` saves `cons` with IV and only one absvar.

## gtools-1.11.6 (2023-11-08)

### Bug fixes

- `greg` with `glm` and absorb no longer gives error (can't dereference NULL)

## gtools-1.11.5 (2023-10-27)

### Bug fixes

- `greg` computes `rss`, `tss` internally with by (previous version would not
  work because the error term was not saved by group, it was overwritten; now
  the full residual vector is saved).

## gtools-1.11.4 (2023-10-24)

### Features

- `greg` saves `rss`, `tss`, `r2` in mata (not with `by()`)
- `greg` saves `cons` in mata with absorb (not with `by()`)

## gtools-1.11.3 (2023-09-20)

### Bug fixes

- Allow `var`, `sd`, `semean`, `cv` with `aw` and weights that add up to 1
  (previously if weights added up to 1 the function exited, but with `aw`
  the bias adjustment is based on the number of observations, not the sum).

## gtools-1.11.2 (2023-08-27)

### Bug fixes

- Allow replace with `fasterxtile` and `gegen xtile`
- `greg` `njabsorb` fills correctly with `by` (pointer was not incrementing in fun)

## gtools-1.11.1 (2023-04-03)

### Bug fixes

- `gegen` now accepts quotes in expressions
- `gregress` no longer forces diagonal vcov matrix (!)
- `gregress` gets header when displaying results
- `fasterxtile` now takes init

## gtools-1.11.0 (2023-03-24)

### Features

- `gregress` adds `alphas()` to save individual fixed effects; `predict()`
  gives right prediction across functions.

## gtools-1.10.2 (2023-02-21)

### Features

- `gstats tab` adds `formatOutput()` to get a matrix with outptu formatted.

## gtools-1.10.1 (2022-12-05)

### Bug fixes

- `gstats transform [if] [in], replace nogreedy` not allowed (since nogreedy
  reads groups on the fly it can't initialize the entire variable).

## gtools-1.10.0 (2022-12-04)

### Features

- `noinit` option for `gcollapse, merge`, `gegen`, `gstats` (selected),
  `gregress` (and co.) to prevent targets from being emptied out
  with `replace`. Prints warning!

### Enhancements

- `gtop` prints the number of levels in Other and Missing rows by
  default. (With missing it only does it if there's more than
  one type of missing value.)

- `greshape` tries to detect repeated stubs and suggests this possibility
  to the user when a stub matches multiple variables.

### Bug fixes

- Closes #87: For OSX, make now compiles x86_64 and arm64 separately 
  then combines via `lipo`.

- `geomean` (`gcollapse`, `gegen`, `gstats`) no longer exhibits
  inconsistent behavior with zeros and negative values. A negative value
  results in a missing value evne if there are zeros present (internally, 
  if a zero is encountered the loop just checks for negative values and
  returns 0 if it finds none).

- `gstats winsor`, exits with error if replace and if/in are passed
  (the way it's set up it'd be a bit of a hassle to allow init/noinit).

- `gstats transform`, `gstats hdfe`, `gregress` (and co.) all now 
  initialize their targets to be empty (missing values) with
  if in and replace.

- `gstats transform` now exits with error if excludeself is passed
  without range, and now prints a warning if it's passed with
  stats other than range.

- `gtop` no longer incorrectly replaces the display value if the
  numerical variable has a value label and no missing values. If there
  was a single value this would result in an error: `gtop` would think
  there was always at least one missing value to replace.

- `gcollapse` no longer fails when trying to label the collapsed output
  if the source labels are blank (this can happen for example with data
  transformed to `.dta` from other formats or programs).

- `gcollapse` no longer gives incorrect missing variables list when
  part of that list is called with varlist notation (e.g. `x* y`
  and `x*` exist but `y` does not).

- `gunique` no longer ignores if/in with `gen` and `replace`

## gtools-1.9.2 (2022-09-22)

### Bug fixes

- Fixed incorrect results with excludeself without specified range in `gstats transform`
  (if there were missing values the input buffer was not being copied, but replaced).

### Features

- Faster excludeself mean and sum without specified range in `gstats transform`.

## gtools-1.9.1 (2022-03-28)

### Features

- Adds various algorithms for projection (cg, squarem, it, map) to
  `gstats hdfe`, `gregress`, `givregress`, `gglm`.
- Some ancillary options (`tolerance()`, `traceiter`, `standardize`)
- Parallel execution of select functions can be enabled at compile
  time via `GTOOLSOMP`
- Typed (direct/non-hashed) radix sort to API internals

## gtools-1.9.0 (2022-03-14)

### Features

- Add `gstats hdfe` (alias `gstats residualize`) for residualizing
  variables (i.e. HDFE transform).

## gtools-1.8.4 (2022-02-02)

### Enhancements

- Move plugin compilation to github rather than travis.

### Bug Fixes

- `gregress` and accompanying commands were not parsing absorb variable types
  since version 1.8.2! The relevant code had been commented out. In effect
  all variables were being taken to be integers.
- `gregress` and accompanying commands with HDFE no longer exit prematurely if
  there are redundant absorbed FEs next to each other.
- Closes #85: Bug in `gegen` warning message causes errors in some fun calls.
- Closes #84, #83, #73: OSX compilation moved to github.

## gtools-1.8.3 (2021-11-03)

### Bug Fixes

- Closes #82: `cw` in `gcollapse` now working.

## gtools-1.8.2 (2021-08-29)

### Features

- Add option to save residuals and predict (the behavior of `predict()` is not
  consistent at the moment; use with extreme caution).

### Bug Fixes

- Closes #79: Adds disclaimer to benchmarks.

## gtools-1.8.1 (2021-08-27)

### Bug Fixes

- Closes #78: if now passed raw/in double-quotes throughout the pipeline
- Closes #75: gunique returns 0s in r() when there are no obs
- Closes #74: gstats transform parses abbreviated targets
- Closes #72: Warning for gegen expressions without by group

build.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# ---------------------------------------------------------------------
# Program: build.py
# Author:  Mauricio Caceres Bravo <mauricio.caceres.bravo@gmail.com>
# Created: Sun Oct 15 10:26:39 EDT 2017
# Updated: Mon Dec 05 09:41:29 EST 2022
# Purpose: Main build file for gtools (copies contents into ./build and
#          puts a .zip file in ./releases)

from os import makedirs, path, linesep, chdir, system, remove, rename
from shutil import copy2, rmtree
from sys import platform
from tempfile import gettempdir
from zipfile import ZipFile
from re import search
import argparse

# ---------------------------------------------------------------------
# Aux programs

try:
    from shutil import which
except:
    def which(program):
        import os

        def is_exe(fpath):
            return path.isfile(fpath) and os.access(fpath, os.X_OK)

        fpath, fname = path.split(program)
        if fpath:
            if is_exe(program):
                return program
        else:
            for epath in os.environ["PATH"].split(os.pathsep):
                epath = epath.strip('"')
                exe_file = path.join(epath, program)
                if is_exe(exe_file):
                    return exe_file

        return None


def makedirs_safe(directory):
    try:
        makedirs(directory)
        return directory
    except OSError:
        if not path.isdir(directory):
            raise


# ---------------------------------------------------------------------
# Command line parsing

parser = argparse.ArgumentParser()
parser.add_argument('--stata',
                    nargs    = 1,
                    type     = str,
                    metavar  = 'STATA',
                    default  = None,
                    required = False,
                    help     = "Path to stata executable")
parser.add_argument('--stata-args',
                    nargs    = 1,
                    type     = str,
                    metavar  = 'STATA_ARGS',
                    default  = None,
                    required = False,
                    help     = "Arguments to pass to Stata executable")
parser.add_argument('--make-flags',
                    nargs    = 1,
                    type     = str,
                    metavar  = 'MAKE_FLAGS',
                    default  = None,
                    required = False,
                    help     = "Arguments to pass to make")
parser.add_argument('--clean',
                    dest     = 'clean',
                    action   = 'store_true',
                    help     = "Clean build",
                    required = False)
parser.add_argument('--no-compile',
                    dest     = 'no_compile',
                    action   = 'store_false',
                    help     = "do not re-compile",
                    required = False)
parser.add_argument('--replace',
                    dest     = 'replace',
                    action   = 'store_true',
                    help     = "Replace build",
                    required = False)
parser.add_argument('--test',
                    dest     = 'test',
                    action   = 'store_true',
                    help     = "Run tests",
                    required = False)
args = vars(parser.parse_args())

# ---------------------------------------------------------------------
# Relevant files

gtools_ssc = [
    "_gtools_internal.ado",
    "_gtools_internal.mata",
    "lgtools.mlib",
    "gcollapse.ado",
    "gcontract.ado",
    "gegen.ado",
    "gunique.ado",
    "gdistinct.ado",
    "glevelsof.ado",
    "gtop.ado",
    "gtoplevelsof.ado",
    "gisid.ado",
    "greshape.ado",
    "greg.ado",
    "gregress.ado",
    "givregress.ado",
    "gglm.ado",
    "gpoisson.ado",
    "gstats.ado",
    "gduplicates.ado",
    "gquantiles.ado",
    "fasterxtile.ado",
    "hashsort.ado",
    "gtools.ado",
    "gcollapse.sthlp",
    "gcontract.sthlp",
    "gegen.sthlp",
    "gunique.sthlp",
    "gdistinct.sthlp",
    "glevelsof.sthlp",
    "gtop.sthlp",
    "gtoplevelsof.sthlp",
    "gisid.sthlp",
    "greshape.sthlp",
    "greg.sthlp",
    "gregress.sthlp",
    "givregress.sthlp",
    "gglm.sthlp",
    "gstats.sthlp",
    "gstats_transform.sthlp",
    "gstats_range.sthlp",
    "gstats_moving.sthlp",
    "gstats_winsor.sthlp",
    "gstats_residualize.sthlp",
    "gstats_hdfe.sthlp",
    "gstats_tab.sthlp",
    "gstats_sum.sthlp",
    "gstats_summarize.sthlp",
    "gduplicates.sthlp",
    "gquantiles.sthlp",
    "fasterxtile.sthlp",
    "hashsort.sthlp",
    "gtools.sthlp",
]

gtools_zip = [
    "changelog.md",
    "gtools.pkg",
    "stata.toc",
] + gtools_ssc

gtools_build = gtools_zip + [
    "gtools_tests.do"
]

# ---------------------------------------------------------------------
# Run the script

# Remove buld
# -----------

rc = 0
if args['clean']:
    print("Removing build files")
    for bfile in gtools_build:
        try:
            remove(path.join("build", bfile))
            print("\tdeleted " + bfile)
        except:
            try:
                remove(path.join("build", "gtools", bfile))
                print("\tdeleted " + bfile)
            except:
                print("\t" + bfile + " not found")

    if args['no_compile']:
        rc = system("make clean SPI=2.0 SPIVER=v2")
        rc = system("make clean SPI=3.0 SPIVER=v3")

    exit(0)

makedirs_safe(path.join("build", "gtools"))
makedirs_safe("releases")

# Stata executable
# ----------------

# I don't have stata on my global path, so to make the script portable
# I make it look for my local executable when Stata is not found.
if args['stata'] is not None:
    statadir = path.abspath(".")
    stataexe = args['stata'][0]
    statargs = "-b do" if args['stata_args'] is None else args['stata_args'][0]
    statado  = '"{0}" {1}'.format(stataexe, statargs)
elif which("stata") is None:
    statadir = path.expanduser("~/.local/stata13")
    stataexe = path.join(statadir, "stata")
    statargs = "-b do" if args['stata_args'] is None else args['stata_args']
    statado  = '"{0}" {1}'.format(stataexe, statargs)
else:
    statadir = path.abspath(".")
    stataexe = 'stata'
    statargs = "-b do" if args['stata_args'] is None else args['stata_args']
    statado  = '"{0}" {1}'.format(stataexe, statargs)

# Temporary files
# ---------------

maindir   = path.dirname(path.realpath(__file__))
tmpdir    = gettempdir()
tmpfile   = path.join(tmpdir, ".compile_lgtools.do")
tmpupdate = path.join(tmpdir, ".update_gtools.do")

# Compile mlib files
# ------------------

matafiles = [path.join("src", "ado", "_gtools_internal.mata")]
with open(path.join("build", "gtools", "gtools.mata"), 'w') as outfile:
    for mfile in matafiles:
        with open(mfile) as infile:
            outfile.write(infile.read())


if which(stataexe):
    with open(tmpfile, 'w') as f:
        f.write("global maindir {0}".format(maindir))
        f.write(linesep)
        f.write("mata: mata set matastrict on")
        f.write(linesep)
        f.write("mata: mata set mataoptimize on")
        f.write(linesep)
        f.write('cd "${maindir}/build/gtools"')
        f.write(linesep)
        f.write("do gtools.mata")
        f.write(linesep)
        f.write("mata")
        f.write(linesep)
        f.write('mata mlib create lgtools, dir("${maindir}/build/gtools") replace')
        f.write(linesep)
        f.write("mata mlib add lgtools Gtools*()")
        f.write(linesep)
        f.write("end")
        f.write(linesep)

    chdir(statadir)
    system(statado + " " + tmpfile)
    print("Compiled lgtools.mlib")
    chdir(maindir)

    copy2(
        path.join("build", "gtools", "lgtools.mlib"),
        path.join("lib",   "plugin", "lgtools.mlib")
    )
else:
    copy2(
        path.join("lib",   "plugin", "lgtools.mlib"),
        path.join("build", "gtools", "lgtools.mlib")
    )

if not path.isfile(path.join("build", "gtools", "lgtools.mlib")):
    print("ERROR: Failed to compile build/gtools/lgtools.mlib")
    exit(-1)
else:
    print("Found build/gtools/lgtools.mlib")

print("")


# Compile plugin files
# --------------------

if args['no_compile']:
    if platform in ["linux", "linux2", "win32", "cygwin", "darwin"]:
        print("Trying to compile plugins for -gtools-")
        make_flags = args['make_flags'][0] if args['make_flags'] is not None else ""
        rc = system("make all SPI=2.0 SPIVER=v2 {0}".format(make_flags))
        rc = system("make all SPI=3.0 SPIVER=v3 {0}".format(make_flags))
        print("Success!" if rc == 0 else "Failed.")
    else:
        print("Don't know platform '{0}'; compile manually.".format(platform))
        exit(198)

print("")

# Get unit test files
# -------------------

testfile = open(path.join("src", "test", "gtools_tests.do")).readlines()
files    = [path.join("src", "test", "test_gcollapse.do"),
            path.join("src", "test", "test_gcontract.do"),
            path.join("src", "test", "test_gquantiles.do"),
            path.join("src", "test", "test_gquantiles_by.do"),
            path.join("src", "test", "test_gegen.do"),
            path.join("src", "test", "test_gunique.do"),
            path.join("src", "test", "test_glevelsof.do"),
            path.join("src", "test", "test_gtoplevelsof.do"),
            path.join("src", "test", "test_gisid.do"),
            path.join("src", "test", "test_greshape.do"),
            path.join("src", "test", "test_gregress.do"),
            path.join("src", "test", "test_gstats.do"),
            path.join("src", "test", "test_gduplicates.do"),
            path.join("src", "test", "test_hashsort.do")]

with open(path.join("build", "gtools_tests.do"), 'w') as outfile:
    outfile.writelines(testfile[:-4])

with open(path.join("build", "gtools_tests.do"), 'a') as outfile:
    for fname in files:
        with open(fname) as infile:
            outfile.write(infile.read())

    outfile.writelines(testfile[-5:])

# Copy files to ./build
# ---------------------

gdir = path.join("build", "gtools")
copy2("changelog.md", gdir)

copy2(path.join("src", "gtools.pkg"),         gdir)
copy2(path.join("src", "stata.toc"),          gdir)

copy2(
    path.join("docs", "stata", "gtoplevelsof.sthlp"),
    path.join("docs", "stata", "gtop.sthlp")
)

copy2(
    path.join("docs", "stata", "gstats_hdfe.sthlp"),
    path.join("docs", "stata", "gstats_residualize.sthlp")
)

copy2(
    path.join("docs", "stata", "gquantiles.sthlp"),
    path.join("docs", "stata", "fasterxtile.sthlp")
)

copy2(
    path.join("docs", "stata", "gregress.sthlp"),
    path.join("docs", "stata", "greg.sthlp")
)

copy2(
    path.join("docs", "stata", "gstats_summarize.sthlp"),
    path.join("docs", "stata", "gstats_sum.sthlp")
)

copy2(
    path.join("docs", "stata", "gstats_summarize.sthlp"),
    path.join("docs", "stata", "gstats_tab.sthlp")
)

copy2(
    path.join("docs", "stata", "gstats_transform.sthlp"),
    path.join("docs", "stata", "gstats_range.sthlp")
)

copy2(
    path.join("docs", "stata", "gstats_transform.sthlp"),
    path.join("docs", "stata", "gstats_moving.sthlp")
)

copy2(path.join("docs", "stata", "gcollapse.sthlp"),          gdir)
copy2(path.join("docs", "stata", "gcontract.sthlp"),          gdir)
copy2(path.join("docs", "stata", "gegen.sthlp"),              gdir)
copy2(path.join("docs", "stata", "gunique.sthlp"),            gdir)
copy2(path.join("docs", "stata", "gdistinct.sthlp"),          gdir)
copy2(path.join("docs", "stata", "glevelsof.sthlp"),          gdir)
copy2(path.join("docs", "stata", "gtop.sthlp"),               gdir)
copy2(path.join("docs", "stata", "gtoplevelsof.sthlp"),       gdir)
copy2(path.join("docs", "stata", "gisid.sthlp"),              gdir)
copy2(path.join("docs", "stata", "greshape.sthlp"),           gdir)
copy2(path.join("docs", "stata", "greg.sthlp"),               gdir)
copy2(path.join("docs", "stata", "gregress.sthlp"),           gdir)
copy2(path.join("docs", "stata", "givregress.sthlp"),         gdir)
copy2(path.join("docs", "stata", "gglm.sthlp"),               gdir)
copy2(path.join("docs", "stata", "gstats.sthlp"),             gdir)
copy2(path.join("docs", "stata", "gstats_transform.sthlp"),   gdir)
copy2(path.join("docs", "stata", "gstats_range.sthlp"),       gdir)
copy2(path.join("docs", "stata", "gstats_moving.sthlp"),      gdir)
copy2(path.join("docs", "stata", "gstats_winsor.sthlp"),      gdir)
copy2(path.join("docs", "stata", "gstats_residualize.sthlp"), gdir)
copy2(path.join("docs", "stata", "gstats_hdfe.sthlp"),        gdir)
copy2(path.join("docs", "stata", "gstats_summarize.sthlp"),   gdir)
copy2(path.join("docs", "stata", "gstats_sum.sthlp"),         gdir)
copy2(path.join("docs", "stata", "gstats_tab.sthlp"),         gdir)
copy2(path.join("docs", "stata", "gduplicates.sthlp"),        gdir)
copy2(path.join("docs", "stata", "gquantiles.sthlp"),         gdir)
copy2(path.join("docs", "stata", "fasterxtile.sthlp"),        gdir)
copy2(path.join("docs", "stata", "hashsort.sthlp"),           gdir)
copy2(path.join("docs", "stata", "gtools.sthlp"),             gdir)

copy2(path.join("src", "ado", "_gtools_internal.ado"),  gdir)
copy2(path.join("src", "ado", "_gtools_internal.mata"), gdir)
copy2(path.join("src", "ado", "gcollapse.ado"),         gdir)
copy2(path.join("src", "ado", "gcontract.ado"),         gdir)
copy2(path.join("src", "ado", "gegen.ado"),             gdir)
copy2(path.join("src", "ado", "gunique.ado"),           gdir)
copy2(path.join("src", "ado", "gdistinct.ado"),         gdir)
copy2(path.join("src", "ado", "glevelsof.ado"),         gdir)
copy2(path.join("src", "ado", "gtop.ado"),              gdir)
copy2(path.join("src", "ado", "gtoplevelsof.ado"),      gdir)
copy2(path.join("src", "ado", "gisid.ado"),             gdir)
copy2(path.join("src", "ado", "greshape.ado"),          gdir)
copy2(path.join("src", "ado", "greg.ado"),              gdir)
copy2(path.join("src", "ado", "gregress.ado"),          gdir)
copy2(path.join("src", "ado", "givregress.ado"),        gdir)
copy2(path.join("src", "ado", "gglm.ado"),              gdir)
copy2(path.join("src", "ado", "gpoisson.ado"),          gdir)
copy2(path.join("src", "ado", "gstats.ado"),            gdir)
copy2(path.join("src", "ado", "gduplicates.ado"),       gdir)
copy2(path.join("src", "ado", "gquantiles.ado"),        gdir)
copy2(path.join("src", "ado", "fasterxtile.ado"),       gdir)
copy2(path.join("src", "ado", "hashsort.ado"),          gdir)
copy2(path.join("src", "ado", "gtools.ado"),            gdir)

# Copy files to .zip folder in ./releases
# ---------------------------------------

# Get stata version
with open(path.join("src", "ado", "gtools.ado"), 'r') as f:
    line    = f.readline()
    version = search('(\d+\.?)+', line).group(0)

plugins = [
    "gtools_unix_v2.plugin",
    "gtools_windows_v2.plugin",
    "gtools_macosx_v2.plugin",
    "gtools_unix_v3.plugin",
    "gtools_windows_v3.plugin",
    "gtools_macosx_v3.plugin"
]

plugbak = plugins[:]
for plug in plugbak:
    if not path.isfile(path.join("build", plug)):
        alt = path.join("lib", "plugin", plug)
        if path.isfile(alt):
            copy2(alt, "build")
        else:
            print("Could not find '{0}'".format(plug))

chdir("build")
print("Compressing build files for gtools-{0}".format(version))
if rc == 0:
    gtools_anyplug = False
    for plug in plugbak:
        if path.isfile(plug):
            gtools_anyplug = True
            rename(path.join(plug), path.join("gtools", plug))
        else:
            plugins.remove(plug)
            print("\t'{0}' not found; skipping.".format(plug))

    if not gtools_anyplug:
        print("WARNING: Could not find plugins despite build exit with 0 status.")
        exit(-1)

    gtools_zip += plugins
else:
    print("WARNING: Failed to build plugins. Will exit.")
    exit(-1)

outzip = path.join(maindir, "releases", "gtools-latest.zip".format(version))
with ZipFile(outzip, 'w') as zf:
    for zfile in gtools_zip:
        zf.write(path.join("gtools", zfile))
        print("\t" + path.join("gtools", zfile))
        rename(path.join("gtools", zfile), zfile)

chdir(maindir)
rmtree(path.join("build", "gtools"))

# Copy files to send to SSC
# -------------------------

print("")
print("Compressing build files for gtools-ssc.zip")
if rc == 0:
    gtools_ssc += plugins
else:
    print("WARNING: Failed to build plugins. Will exit.")
    exit(-1)

chdir("build")
outzip = path.join(maindir, "releases", "gtools-ssc.zip")
with ZipFile(outzip, 'w') as zf:
    for zfile in gtools_ssc:
        zf.write(zfile)
        print("\t" + zfile)

# Replace package in ~/ado/plus
# -----------------------------

chdir(maindir)
if args["replace"]:
    if which(stataexe):
        with open(tmpupdate, 'w') as f:
            f.write("global builddir {0}".format(path.join(maindir, "build")))
            f.write(linesep)
            f.write("cap net uninstall gtools")
            f.write(linesep)
            f.write("net install gtools, from($builddir) replace")
            f.write(linesep)

        chdir(statadir)
        system(statado + " " + tmpupdate)
        remove(tmpupdate)
        # print(linesep + "Replaced gtools in ~/ado/plus")
        chdir(maindir)
    else:
        print("Could not find Stata executable '{0}'.".format(stataexe))
        exit(-1)

# Run tests
# ---------

if args['test']:
    print("Running tests (see build/gtools_tests.log for output)")
    chdir("build")
    system(statado + " gtools_tests.do")
    chdir(maindir)
